File: LPIS_1/src/main/java/com/example/lab/statement/StatementContext.java
package com.example.lab.statement;

import com.example.lab.*;

import java.util.Optional;

public interface StatementContext extends FunctionContext {

    StatementContext funcDefBlock();

    StatementContext forDefBlock();

    StatementContext block();

    boolean isReturn();

    void setReturn();

    boolean isFuncDef();

    boolean isForDef();

    void funcDef(Function function);

    Optional<ReturnType> getReturnType(FunctionSignature function);

    Optional<Type> getVariableType(String variable);

    boolean hasFunction(FunctionSignature signature);

    default boolean hasVariable(Variable variable) {
        return hasVariable(variable.name());
    }

    boolean hasVariable(String name);

    default void initVariable(String name) {
        initVariable(new Variable(name));
    }

    void initVariable(Variable variable);

}

File: LPIS_1/src/main/java/com/example/lab/statement/StatementContext.kt
package com.example.lab.statement

import com.example.lab.Function
import com.example.lab.FunctionContext
import com.example.lab.FunctionSignature
import com.example.lab.Variable

interface StatementContext : FunctionContext {

	fun funcDefBlock(): StatementContext

	fun forDefBlock(): StatementContext

	override fun block(): StatementContext

	val isReturn: Boolean

	fun setReturn()

	val isFuncDef: Boolean
	val isForDef: Boolean

	fun funcDef(function: Function)

	fun hasFunction(signature: FunctionSignature): Boolean

	fun hasVariable(variable: Variable): Boolean {
		return hasVariable(variable.name)
	}

	fun hasVariable(name: String): Boolean

	fun initVariable(name: String) {
		initVariable(Variable(name))
	}

	fun initVariable(variable: Variable)
}
